.github/workflows/build.yaml

```diff
@@ -0,0 +1,200 @@
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        include:
+        - odoo_serie: "11.0"
+        - odoo_serie: "12.0"
+        - odoo_serie: "13.0"
+        - odoo_serie: "14.0"
+        - odoo_serie: "15.0"
+        - odoo_serie: "16.0"
+    steps:
+
+
+      - uses: actions/download-artifact@v2
+        with:
+          name: test-compose.yml
+
+      - name: Download the image from artifact
+        uses: ishworkh/docker-image-artifact-download@v1
+        with:
+          image: "${{ matrix.odoo_serie }}:latest"
+      - name: Run tests
+        run: export BASE_IMAGE=${{ matrix.odoo_serie }}:latest ; docker-compose -f test-compose.yml run --rm -e LOCAL_USER_ID=999 $@ odoo runtests
+
+  push:
+    needs: test
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        include:
+        - odoo_serie: "11.0"
+        - odoo_serie: "12.0"
+        - odoo_serie: "13.0"
+        - odoo_serie: "14.0"
+        - odoo_serie: "15.0"
+        - odoo_serie: "16.0"
+
+    if: github.event_name == 'push'
+
+
+
+    steps:
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v1
+
+
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v4
+        with:
+          images: ghcr.io/camptocamp/docker-odoo-project
+          flavor: |
+              prefix=${{ matrix.odoo_serie }}-,onlatest=true
+          tags: |
+              type=raw,value={{matrix.odoo_serie}}latest
+              type=raw,value={{matrix.odoo_serie}}{{date 'YYYYMMDD'}}
+              type=ref,event=tag
+
+
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ secrets.GHCR_USER }}
+          password: ${{ secrets.GHCR_TOKEN }}
+
+      - name: Prepare src
+        run: |
+          cp -rT common/ ${{ matrix.odoo_serie }}_dir
+          cp -r ${{ matrix.odoo_serie }}/. ${{ matrix.odoo_serie }}_dir
+          cp -r bin/ ${{ matrix.odoo_serie }}_dir
+          cp -r bin-py2/ ${{ matrix.odoo_serie }}_dir
+          cp -rT common/ ${{ matrix.odoo_serie }}_dir
+          cp ${{ matrix.odoo_serie }}_dir/Dockerfile-onbuild ${{ matrix.odoo_serie }}_dir/Dockerfile-batteries-onbuild
+          sed -i "1i FROM ${BUILD_TAG}" ${{ matrix.odoo_serie }}_dir/Dockerfile-onbuild
+          sed -i "1i FROM ${BUILD_TAG}" ${{ matrix.odoo_serie }}_dir/Dockerfile-batteries
+          sed -i "1i FROM ${BUILD_TAG}-batteries" ${{ matrix.odoo_serie }}_dir/Dockerfile-batteries-onbuild
+          cp -r install/ ${{ matrix.odoo_serie }}_dir
+          cp -r start-entrypoint.d/ ${{ matrix.odoo_serie }}_dir
+          cp -r before-migrate-entrypoint.d/ ${{ matrix.odoo_serie }}_dir
+
+
+      - name: Build
+        uses: docker/build-push-action@v3
+        with:
+          context: ./${{ matrix.odoo_serie }}_dir
+          push: ${{ github.event_name != 'pull_request' }}
+          tags: ${{ steps.meta.outputs.tags }}
```

Comment:
Instead of building it a second time, it would be more efficient to download the image we pushed into artifacts.

Reply:
Or use a single job that builds, tests and push in multiple steps.
It would be even faster as it would run on the same vm

Reply:
Indeed, AFAIK this was copied from an other github action where the middle phase is done in parallel for 2 testing vm.